### PR TITLE
KAFKA-3561: Auto create through topic for KStream aggregation and join [WIP]

### PR DESCRIPTION
**Pull request by dguy**
@guozhangwang can you please take a look at this? It is not close to being done but i'd like some feedback to ensure i'm heading down the right path and understand what this JIRA is asking. The main things to look at are the `KStreamImpl.map(...)` method I added and the change made to `TopologyBuilder.copartitionGroups`

There is also a test, `KStreamRepartitionMappedKeyTest`, that passes with these changes.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java [hunk 1]
```diff
@@ -620,22 +621,35 @@ public final TopologyBuilder copartitionSources(Collection<String> sourceNodes)
      * Returns the copartition groups.
      * A copartition group is a group of source topics that are required to be copartitioned.
      *
+     * @param applicationId
      * @return groups of topic names
      */
-    public Collection<Set<String>> copartitionGroups() {
+    public Collection<Set<String>> copartitionGroups(final String applicationId) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:627]: Is this just for code cleanness or is necessary for correctness?

**Comment by guozhangwang** [comment]
Hi @dguy , thanks for the patch. Overall looks good. Here is one general comment:

We should not require serdes in `KStream.map()` as it is supposed to be non-stateful operators and hence users are not expected to have any materialization JUST FOR this operator. Plus, some other operators like `selectKey` would also need similar handing if we do this direction.

As about join operations, since the join operators already requires the serdes which would be just reused for the repartition topic, we do not need to require it before hand in `map`, `selectKeys`, etc. Similar situation for `reduce`, which already requires the serdes for original key and value.

Hence the only problem is for `aggregate`, where the original value serde is not provided. After thinking about it for a bit more, I'm now more towards doing KAFKA-3576 for aggregations as well, i.e. introduce a `KGroupedStream` which can be generated from `KStream.groupBy` (which can then require serdes). In addition, `KStream.groupBy` may also take a `Windows` spec (we can overload the function with and without this parameter) so that later aggregations do not need to specify it anymore.

**Comment by guozhangwang** [comment]
A few more suggestions:
1. As in `KGroupedTable KTable.groupBy(..)`, the repartitioning does not happen at the `groupBy` operator, but we carry the serde information as well as the select-key mapper along with the `KGroupedTable`, which can only call aggregations; then the repartitioning is done when `aggregate / reduce / etc` is really called. We can do the same for `KStream`.
2. A few differences of the `KGroupedStream` though, is that besides the serde and key-select mapper information, we can also allow users to specify the Windows spec so that the `KGroupedStream` conceptually captures a stream grouped by key and as well as windows (this is from user interfacing point of view only, no correctness issues at all), then its aggregation function do not need to provide any Windowing specs any more.
3. Also for `KGroupedStream.reduce`, we do not need to require users provide any more serdes, as they should already be provided in the previous `groupBy`.
4. As for joins, we are going to still let users do the pattern of `stream1.selectKey().join(stream2, serdes..)` or `stream1.map().join(stream2, serdes..)`, but auto-creating the repartition topic inside the `join` operator; note that again since the serdes are passed in the later join operator, we do not need to enforce users to provide them in previous `selectKey` or `map` operator.

Let me know if you have more questions.

**Comment by dguy** [comment]
Thanks for the clarification

**Pull request closed without merging**